src/Guard.php

```diff
@@ -67,146 +76,127 @@ class Guard
      * false to get a new token with each request.
      */
     protected $persistentTokenMode;
-    
+
     /**
      * Stores the latest key-pair generated by the class
      *
-     * @var array
+     * @var array|null
      */
     protected $keyPair;
 
     /**
-     * Create new CSRF guard
-     *
-     * @param string                 $prefix
-     * @param null|array|ArrayAccess $storage
-     * @param null|callable          $failureCallable
-     * @param integer                $storageLimit
-     * @param integer                $strength
-     * @param boolean                $persistentTokenMode
+     * @param ResponseFactoryInterface  $responseFactory
+     * @param string                    $prefix
+     * @param null|array|ArrayAccess    $storage
+     * @param null|callable             $failureHandler
+     * @param integer                   $storageLimit
+     * @param integer                   $strength
+     * @param boolean                   $persistentTokenMode
      * @throws RuntimeException if the session cannot be found
      */
     public function __construct(
-        $prefix = 'csrf',
+        ResponseFactoryInterface $responseFactory,
+        string $prefix = 'csrf',
         &$storage = null,
-        callable $failureCallable = null,
-        $storageLimit = 200,
-        $strength = 16,
-        $persistentTokenMode = false
+        callable $failureHandler = null,
+        int $storageLimit = 200,
+        int $strength = 16,
+        bool $persistentTokenMode = false
     ) {
+        $this->responseFactory = $responseFactory;
         $this->prefix = rtrim($prefix, '_');
+
         if ($strength < 16) {
-            throw new RuntimeException('CSRF middleware failed. Minimum strength is 16.');
+            throw new RuntimeException('CSRF middleware instantiation failed. Minimum strength is 16.');
         }
         $this->strength = $strength;
-        $this->storage = &$storage;
 
-        $this->setFailureCallable($failureCallable);
+        $this->setStorage($storage);
+        $this->setFailureHandler($failureHandler);
         $this->setStorageLimit($storageLimit);
-
         $this->setPersistentTokenMode($persistentTokenMode);
 
         $this->keyPair = null;
     }
 
     /**
-     * Retrieve token name key
-     *
-     * @return string
+     * @param null|array|ArrayAccess
+     * @throws RuntimeException
      */
-    public function getTokenNameKey()
+    public function setStorage(&$storage = null): void
     {
-        return $this->prefix . '_name';
+        if (is_array($storage) || ($storage instanceof ArrayAccess)) {
+            $this->storage = &$storage;
+            return;
+        }
+
+        if (!isset($_SESSION)) {
```

Comment:
The variable `$_SESSION` might be set if the session had been started and destroyed (in most cases, this would not really happen, but anyway). Probably we should use the following if condition:
```php
        if (session_status() !== PHP_SESSION_ACTIVE) {
```